be/src/formats/parquet/level_builder.h

```diff
@@ -0,0 +1,153 @@
+
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <arrow/api.h>
+#include <arrow/buffer.h>
+#include <arrow/io/api.h>
+#include <arrow/io/file.h>
+#include <arrow/io/interfaces.h>
+#include <gen_cpp/DataSinks_types.h>
+#include <parquet/api/reader.h>
+#include <parquet/api/writer.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/exception.h>
+
+#include <utility>
+
+#include "column/chunk.h"
+#include "column/nullable_column.h"
+#include "formats/parquet/chunk_writer.h"
+#include "fs/fs.h"
+#include "runtime/runtime_state.h"
+#include "util/priority_thread_pool.hpp"
+
+namespace starrocks::parquet {
+
+// Intermediate data passed between add_column_chunk functions.
+// Immutable and thread-safe.
+class LevelBuilderContext {
+public:
+    LevelBuilderContext(size_t num_levels, std::shared_ptr<std::vector<int16_t>> def_levels = nullptr,
+                        int16_t max_def_level = 0, std::shared_ptr<std::vector<int16_t>> rep_levels = nullptr,
+                        int16_t max_rep_level = 0)
+            : _max_def_level(max_def_level),
+              _max_rep_level(max_rep_level),
+              _def_levels(std::move(def_levels)),
+              _rep_levels(std::move(rep_levels)),
+              _num_levels(num_levels) {
+        DCHECK(_max_def_level == 0 || _def_levels != nullptr);
+        DCHECK(_max_rep_level == 0 || _rep_levels != nullptr);
+        DCHECK(_max_def_level == 0 || _num_levels == _def_levels->size());
+        DCHECK(_max_rep_level == 0 || _num_levels == _rep_levels->size());
+    }
+
+    int num_levels() const { return _num_levels; }
+
+public:
```

Comment:
do not set it as public

Reply:
I think this is ok because the instantiated objects are immutable. 

Will make this class as a struct if necessary.